src/utils/hooks/useToggle.tsx

```diff
@@ -0,0 +1,15 @@
+import { useReducer } from 'react';
+
+export type DispatchWithOptionalAction<Type> = (_arg?: Type | unknown) => void;
+
+export type UseToggleType<Type> = [Type, DispatchWithOptionalAction<Type>];
+
+function toggle(currentValue: boolean, newValue: boolean | undefined) {
+    return typeof newValue === 'boolean' ? newValue : !currentValue;
+}
+
+function useToggle(initialValue = false): UseToggleType<boolean> {
+    return useReducer(toggle, initialValue);
```

Comment:
I find this implementation cleaner than having to wrap the setter function in another abstraction. Moreover, we can declare the function outside instead of having to redeclare every time. 

Why do you think it adds extra complexity? 
I do agree that it might be more complicated to read as we are probably less used to using `useReducer` than a simple `useState`.

Reply:
but if you return the reducer, it means, we return [dispatch, state]. What usage will be here?